File: templates/Virtual-Assistant-Template/typescript/generator-botbuilder-assistant/generators/app/templates/customAssistant/src/dialogs/onboardingDialog.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */
import {
    BotTelemetryClient,
    StatePropertyAccessor,
    TurnContext } from 'botbuilder';
import {
    ComponentDialog,
    DialogTurnResult,
    TextPrompt,
    WaterfallDialog,
    WaterfallStepContext } from 'botbuilder-dialogs';
import { IOnboardingState } from '../models/onboardingState';
import { OnboardingResponses } from '../responses/onboardingResponses';
import { BotServices } from '../services/botServices';

enum DialogIds {
    namePrompt = 'namePrompt',
    emailPrompt = 'emailPrompt',
    locationPrompt =  'locationPrompt'
}

export class OnboardingDialog extends ComponentDialog {

    // Fields
    private static readonly responder: OnboardingResponses = new OnboardingResponses();
    private readonly accessor: StatePropertyAccessor<IOnboardingState>;
    private state!: IOnboardingState;

    // Constructor
    constructor(botServices: BotServices, accessor: StatePropertyAccessor<IOnboardingState>, telemetryClient: BotTelemetryClient) {
        super(OnboardingDialog.name);
        this.accessor = accessor;
        this.initialDialogId = OnboardingDialog.name;
        const onboarding: ((sc: WaterfallStepContext<IOnboardingState>) => Promise<DialogTurnResult>)[] = [
            this.askForName.bind(this),
            this.finishOnboardingDialog.bind(this)
        ];

        // To capture built-in waterfall dialog telemetry, set the telemetry client
        // to the new waterfall dialog and add it to the component dialog
        this.telemetryClient = telemetryClient;
        this.addDialog(new WaterfallDialog(this.initialDialogId, onboarding));
        this.addDialog(new TextPrompt(DialogIds.namePrompt));
    }

    public async askForName(sc: WaterfallStepContext<IOnboardingState>): Promise<DialogTurnResult> {
        this.state = await this.getStateFromAccessor(sc.context);

        if (this.state.name !== undefined && this.state.name.trim().length > 0) {
            return sc.next(this.state.name);
        }

        return sc.prompt(DialogIds.namePrompt, {
            prompt: await OnboardingDialog.responder.renderTemplate(
                sc.context,
                OnboardingResponses.responseIds.namePrompt,
                <string> sc.context.activity.locale)
        });
    }

    public async finishOnboardingDialog(sc: WaterfallStepContext<IOnboardingState>): Promise<DialogTurnResult> {
        this.state = await this.getStateFromAccessor(sc.context);
        this.state.name = <string> sc.result;
        await this.accessor.set(sc.context, this.state);

        await OnboardingDialog.responder.replyWith(
            sc.context,
            OnboardingResponses.responseIds.haveLocationMessage,
            {
                name: this.state.name,
                location: this.state.location
            });

        return sc.endDialog();
    }

    private async getStateFromAccessor(context: TurnContext): Promise<IOnboardingState>  {
        const state: IOnboardingState | undefined = await this.accessor.get(context);
        if (state === undefined) {
            const newState: IOnboardingState = {
                email: '',
                location: '',
                name: ''
            };
            await this.accessor.set(context, newState);

            return newState;
        }

        return state;
    }
 }

File: samples/csharp/assistants/EnterpriseAssistant/EnterpriseAssistant/Dialogs/OnboardingDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using VirtualAssistant.Models;
using VirtualAssistant.Responses.Onboarding;
using VirtualAssistant.Services;

namespace VirtualAssistant.Dialogs
{
    public class OnboardingDialog : ComponentDialog
    {
        private static OnboardingResponses _responder = new OnboardingResponses();
        private IStatePropertyAccessor<OnboardingState> _accessor;
        private OnboardingState _state;

        public OnboardingDialog(
            BotServices botServices,
            UserState userState,
            IBotTelemetryClient telemetryClient)
            : base(nameof(OnboardingDialog))
        {
            _accessor = userState.CreateProperty<OnboardingState>(nameof(OnboardingState));
            InitialDialogId = nameof(OnboardingDialog);

            var onboarding = new WaterfallStep[]
            {
                AskForName,
                FinishOnboardingDialog,
            };

            // To capture built-in waterfall dialog telemetry, set the telemetry client
            // to the new waterfall dialog and add it to the component dialog
            TelemetryClient = telemetryClient;
            AddDialog(new WaterfallDialog(InitialDialogId, onboarding) { TelemetryClient = telemetryClient });
            AddDialog(new TextPrompt(DialogIds.NamePrompt));
        }

        public async Task<DialogTurnResult> AskForName(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            _state = await _accessor.GetAsync(sc.Context, () => new OnboardingState());

            if (!string.IsNullOrEmpty(_state.Name))
            {
                return await sc.NextAsync(_state.Name);
            }
            else
            {
                return await sc.PromptAsync(DialogIds.NamePrompt, new PromptOptions()
                {
                    Prompt = await _responder.RenderTemplate(sc.Context, sc.Context.Activity.Locale, OnboardingResponses.ResponseIds.NamePrompt),
                });
            }
        }

        public async Task<DialogTurnResult> FinishOnboardingDialog(WaterfallStepContext sc, CancellationToken cancellationToken)
        {
            _state = await _accessor.GetAsync(sc.Context, () => new OnboardingState());
            var name = _state.Name = (string)sc.Result;
            await _accessor.SetAsync(sc.Context, _state, cancellationToken);

            await _responder.ReplyWith(sc.Context, OnboardingResponses.ResponseIds.HaveNameMessage, new { name });
            return await sc.EndDialogAsync();
        }

        private class DialogIds
        {
            public const string NamePrompt = "namePrompt";
        }
    }
}
